dist/stat_data.py

```diff
@@ -335,6 +339,41 @@ def __init__(self, name, desc, flags=''):
+    CheckpointStat('checkpoint_handle_checked', 'transaction checkpoint data handle operation applied'),
+    CheckpointStat('checkpoint_handle_duration', 'transaction checkpoint most recent duration for gathering all handles (usecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_handle_duration_apply', 'transaction checkpoint most recent duration for gathering applied handles (usecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_handle_duration_skip', 'transaction checkpoint most recent duration for gathering skipped handles (usecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_handle_skipped', 'transaction checkpoint most recent handles skipped'),
+    CheckpointStat('checkpoint_handle_walked', 'transaction checkpoint most recent handles walked'),
+    CheckpointStat('checkpoint_prep_max', 'transaction checkpoint prepare max time (msecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_prep_min', 'transaction checkpoint prepare min time (msecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_prep_recent', 'transaction checkpoint prepare most recent time (msecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_prep_running', 'transaction checkpoint prepare currently running', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_prep_total', 'transaction checkpoint prepare total time (msecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_presync', 'number of handles visited after checkpoint writes complete'),
+    CheckpointStat('checkpoint_scrub_target', 'transaction checkpoint scrub dirty target', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_scrub_time', 'transaction checkpoint scrub time (msecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_skipped', 'transaction checkpoints skipped because database was clean'),
+    CheckpointStat('checkpoint_state', 'last potentially expensive operation attempted', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_stop_stress_active', 'transaction checkpoint stop timing stress active', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_sync', 'number of files synced as part of checkpoint'),
+    CheckpointStat('checkpoint_time_max', 'transaction checkpoint max time (msecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_time_min', 'transaction checkpoint min time (msecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_time_recent', 'transaction checkpoint most recent time (msecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_time_total', 'transaction checkpoint total time (msecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_update_postprocess', 'number of older named checkpoints altered to include new checkpoint'),
+    CheckpointStat('checkpoint_update_prev_named', 'number of older named checkpoints updated during reconciliation'),
+    CheckpointStat('checkpoint_wait_reduce_dirty', 'wait cycles while cache dirty level is decreasing'),
```

Comment:
I'd like if we could add statistics about the pages a checkpoint is visiting. Maybe a counter of internal and leaf pages being visited. Also a counter of the number of pages checkpoint causes to be reconciled. A counter of pages read into the cache by checkpoint.
How many records are written to the history store due to checkpoint reconciliation.

Sorry if any of these already exist. Looking at the ticket, and the set here, I'm not convinced they would show what is happening in that 16 minute period when there is no write I/O generated.

Reply:
That's reasonable, I'll have to think a bit harder about some of them - some of this code lives in other subsystems, and it wouldn't make sense to increment checkpoint stats in some general-purpose traversal function.

Reply:
Thanks for the push here - I had another look and realised that `__wt_sync_file` already had a section dedicated to checkpoint, so there was indeed a sensible place to put these stats. I've added them now (although Github seems to be having issues displaying new commits).

I want to confirm one thing: does `page->read_gen` being `WT_READGEN_WONT_NEED` mean it was a page we read into the cache? I have a feeling it could be set some other way, but I don't know for sure.

Reply:
Resolving this offline